src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -1089,9 +1136,22 @@ void Replayer<I>::handle_remove_mirror_peer_uuid(
     int r, const std::string &snap_id) {
   dout(10) << snap_id << ", r=" << r << dendl;
 
-  if (r < 0) {
+  // ENOENT could be resulting because mirror_peer_uuid was removed before,
+  // may be before a daemon restart.
+  if (r < 0 && r != -ENOENT) {
     derr << "failed to remove mirror_peer_uuid for snap_id: "
          << snap_id  << " : " << cpp_strerror(r) << dendl;
+    return;
+  }
+
+  auto iter = std::find_if(
```

Comment:
This is unsafe as the async callback is modifying m_prune_snap_ids without a lock. This could cause issues because clean_unnecessary_mirror_group_snaps calls remove_mirror_peer_uuid and also accesses m_prune_snap_ids.

Reply:
Ack. Will take a lock here.

Reply:
Since `handle_remove_mirror_peer_uuid` needs to take the lock to modify m_prune_snap_ids, the callback thread will potentially be stuck until load_local_group_snapshots() finishes processing the local snaps and unlocks the lock before calling schedule_load_group_snapshots. Can that take a long time?  Will this affect performance especially if there are multiple calls to remove_peeer_uuid?

Reply:
I didn't noticed it so far, I will keep a watch.